### PR TITLE
Login once when putting a collection of files

### PR DESCRIPTION
**Pull request by iwootten**
See issue https://github.com/ONSdigital/perkin/issues/107

**Changes**

- Add login / logout methods to ftppublisher
- Use login/logout when transferring a collection of files at the start/end of transaction

**How to test**

- This can be tested within the docker-compose setup to determine if it is functionally the same, but ideally we'd want to test under preprod to confirm whether the login overhead has been reduced for each submission.

**Who can test**

Anyone but @iwootten

### CODE REVIEW

File: src/main/java/com/github/onsdigital/perkin/publish/FtpPublisher.java [hunk 1]
```diff
@@ -158,7 +188,7 @@ private void put(InputStream inputStream, String path, String filename) throws I
 
                 log.debug("FTP|logout");
                 ftp.logout();
-                log.debug("FTP|disconnect");
+                log.debug("FTP|logout");
```
Comment (dgunthor) [inline src/main/java/com/github/onsdigital/perkin/publish/FtpPublisher.java]: should be 'disconnect' in the log message?

File: src/main/java/com/github/onsdigital/perkin/publish/FtpPublisher.java [hunk 2]
```diff
@@ -41,15 +45,17 @@ public void publish(List<DataFile> files) throws IOException {
...
 
-            String path = determinPath(file.getPath());
+            String path = determinePath(file.getPath());
             put(inputStream, path, filename);
 
             timer.stopStatus(200);
             Audit.getInstance().increment(timer, file);
         }
+
+        logout();
```
Comment (dgunthor) [inline src/main/java/com/github/onsdigital/perkin/publish/FtpPublisher.java]: i think logout() should be within a finally block - if put throws an IOException then the ftp connection will not be closed.
should add a test to show that when put throws an IOException, ftp connection is closed

**Comment by dgunthor** [comment]
not sure of the value of this change request - we are 'fixing' a performance issue that doesn't yet cause us a problem - this is about to break when I introduce multithreading - can this be left as an open issue for now?

**Comment by iwootten** [comment]
Well, the problem is that the performance requirements are not met - hence the open issue. Can you open up a pull request for your multithreading work to discuss to see if the two can work together?

**Pull request merged**